x-pack/plugins/security_solution/server/lib/detection_engine/prebuilt_rules/logic/prebuilt_rules_importer.ts

```diff
@@ -0,0 +1,129 @@
+import { ensureLatestRulesPackageInstalled } from './ensure_latest_rules_package_installed';
+
+type MaybeRule = RuleToImport | Error;
+export interface IPrebuiltRulesImporter {
+  setup: () => Promise<void>;
+  fetchPrebuiltRuleAssets: ({ rules }: { rules: MaybeRule[] }) => Promise<PrebuiltRuleAsset[]>;
+}
+
+/**
+ *
+ * This object contains logic necessary to import prebuilt rules into the system.
+ */
+export class PrebuiltRulesImporter implements IPrebuiltRulesImporter {
+  private context: SecuritySolutionApiRequestHandlerContext;
+  private config: ConfigType;
+  private ruleAssetsClient: ReturnType<typeof createPrebuiltRuleAssetsClient>;
+  public enabled: boolean;
+  public latestPackagesInstalled: boolean = false;
+
+  constructor({
+    config,
+    savedObjectsClient,
+    context,
+  }: {
+    config: ConfigType;
+    context: SecuritySolutionApiRequestHandlerContext;
+    savedObjectsClient: SavedObjectsClientContract;
+  }) {
+    this.ruleAssetsClient = createPrebuiltRuleAssetsClient(savedObjectsClient);
+    this.context = context;
+    this.config = config;
+    this.enabled = config.experimentalFeatures.prebuiltRulesCustomizationEnabled;
+  }
+
+  /**
+   *
+   * Prepares the system to import prebuilt rules by ensuring the latest rules
+   * package is installed.
+   */
+  public async setup() {
+    if (!this.enabled) {
+      return;
+    }
+
+    await ensureLatestRulesPackageInstalled(this.ruleAssetsClient, this.config, this.context);
+    this.latestPackagesInstalled = true;
+  }
+
+  /**
+   * Retrieves the prebuilt rule assets for the specified rules being imported. If
+   * @param rules - The rules to be imported
+   *
+   * @returns The prebuilt rule assets corresponding to the specified prebuilt
+   * rules, which are used to determine how to import those rules (create vs. update, etc.).
+   */
+  public async fetchPrebuiltRuleAssets({
+    rules,
+  }: {
+    rules: MaybeRule[];
+  }): Promise<PrebuiltRuleAsset[]> {
+    if (!this.enabled) {
+      return [];
+    }
+
+    if (!this.latestPackagesInstalled) {
+      throw new Error(
+        'Prebuilt rule assets cannot be fetched until the latest rules package is installed. Call setup() on this object first.'
+      );
+    }
+
+    const prebuiltRulesToImport = rules.flatMap((rule) => {
+      if (rule instanceof Error || rule.version == null) {
+        return [];
+      }
+      return {
+        rule_id: rule.rule_id,
+        version: rule.version,
+      };
+    });
+
+    return this.ruleAssetsClient.fetchAssetsByVersion(prebuiltRulesToImport);
+  }
+
+  /**
+   * Retrieves the rule IDs (`rule_id`) of installed prebuilt rules matching those
+   * of the specified rules. Can be used to determine whether the rule being
+   * imported is a custom rule or a prebuilt rule.
+   *
+   * @param rules - A list of rules being imported.
+   *
+   * @returns A list of the rule IDs that are installed.
+   *
+   */
```

Comment:
Since `version` doesn't matter in either the input or the output, I'm not sure that mentioning `version` would be useful here. The implementation could be changed to call e.g. `fetchEarliestAssetsByRuleId` without this API changing.